net/httpclient_test.go

```diff
@@ -248,6 +255,89 @@ func TestTransport(t *testing.T) {
 			req:         httptest.NewRequest("GET", "http://example.com/", nil),
 			wantErr:     false,
 		},
+		{
+			name: "With hooks, should have request header and respose changed",
+			options: Options{
+				BeforeSend: func(req *http.Request) {
+					if req != nil {
+						req.Header.Set("X-Foo", "bar")
+					}
+				},
+				AfterResponse: func(rsp *http.Response, err error) {
+					if rsp != nil {
+						rsp.StatusCode = 255
+					}
+				},
+			},
+			req:     httptest.NewRequest("GET", "http://example.com/", nil),
+			wantErr: false,
+			checkRequest: func(req *http.Request) error {
+				if v := req.Header.Get("X-Foo"); v != "bar" {
+					return fmt.Errorf(`failed to patch request want "X-Foo": "bar", but got: %s`, v)
+				}
+				return nil
+			},
+			checkResponse: func(rsp *http.Response) error {
+				if rsp.StatusCode != 255 {
+					return fmt.Errorf("failed to get status code 255, got: %d", rsp.StatusCode)
+				}
+				return nil
+			},
+		},
+		{
+			name: "With hooks and opentracing, should have request header and response changed",
+			options: Options{
+				Tracer: mtracer,
+				BeforeSend: func(req *http.Request) {
+					if req != nil {
+						if span := opentracing.SpanFromContext(req.Context()); span != nil {
+							fmt.Printf("1. req: %p\n", req)
+							span.SetTag(string(ext.PeerService), "my-app")
+							*req = *req.WithContext(opentracing.ContextWithSpan(req.Context(), span))
+							fmt.Printf("2. req: %p\n", req)
+							return
+						}
+					}
+					println("BeforeSend code not executed")
+				},
+				AfterResponse: func(rsp *http.Response, err error) {
+					if rsp != nil {
+						if span := opentracing.SpanFromContext(rsp.Request.Context()); span != nil {
+							fmt.Printf("1. rsp.request: %p\n", rsp.Request)
+							span.SetTag("my.status", 255)
+							*rsp.Request = *rsp.Request.WithContext(opentracing.ContextWithSpan(rsp.Request.Context(), span))
+							fmt.Printf("2. rsp.request: %p\n", rsp.Request)
+							return
+						}
+					}
+					println("AfterResponse code not executed")
+				},
+			},
+			spanName: "myspan",
+			req:      httptest.NewRequest("GET", "http://example.com/", nil),
+			wantErr:  false,
+			checkRequest: func(req *http.Request) error {
+				fmt.Printf("1. checkRequest req: %p\n", req)
+				if span := opentracing.SpanFromContext(req.Context()); span != nil {
```

Comment:
span is nil at the server and client side. Test log:
```
% go test ./net -run "TestTransport/With_hooks_and_opentracing,_should_have_request_header_and_response_changed" -count 1 -v
=== RUN   TestTransport
DO NOT USE IN PRODUCTION
=== RUN   TestTransport/With_hooks_and_opentracing,_should_have_request_header_and_response_changed
2024/07/10 13:01:20 injectSpan
1. req: 0xc0000f5c20
2. req: 0xc0000f5c20
before roundtrip req: 0xc0000f5c20
1. checkRequest req: 0xc00015e000
    httpclient_test.go:368: Failed to check request: failed get span from request
after roundtrip: 0xc0000f5c20
1. rsp.request: 0xc0000f5c20
2. rsp.request: 0xc0000f5c20
after roundtrip on the client
1. checkRequest req: 0xc0000f5200
    httpclient_test.go:396: Failed to check request: failed get span from request
--- FAIL: TestTransport (0.00s)
    --- FAIL: TestTransport/With_hooks_and_opentracing,_should_have_request_header_and_response_changed (0.00s)
FAIL
FAIL    github.com/zalando/skipper/net  0.180s
FAIL
zsh: exit 1     go test ./net -run  -count 1 -v
```
see that checkRequest() sees always a different pointer for the modified request than everyone else.

Reply:
Yes but it should!?
Also the last log just before the 2nd error `1. checkRequest req: 0xc0000f5200` on the client, that use `rsp.Request` is not the same as the request that we sent.

Reply:
Span comes from the context and I think there is nothing in test server that adds a span to the context.

Reply:
fixed test by running on the client after roundtrip